File: ios/ThinkingAnalyticsSDK/ThinkingPluginProxy.m
//
//  ThinkingPluginProxy.m
//  App
//
//  Copyright (c) 2012 ThinkingData. All rights reserved.
//  Created by huangdiao on 2022/9/2.
//

#import "ThinkingPluginProxy.h"

#if __has_include(<ThinkingSDK/TDEventModel.h>)
#import <ThinkingSDK/ThinkingAnalyticsSDK.h>
#else
#import "ThinkingAnalyticsSDK.h"
#endif

@implementation ThinkingPluginProxy

+ (ThinkingAnalyticsSDK *)sharedInstance:(NSString *)appId {
    ThinkingAnalyticsSDK * instance = nil;
    if (appId != nil && [appId length] > 0) {
        instance = [ThinkingAnalyticsSDK sharedInstanceWithAppid:appId];
    }
    if (instance == nil) {
        instance = [ThinkingAnalyticsSDK sharedInstance];
    }
    return instance;
}

+ (void)shareInstance:(NSDictionary *)params {
    NSString *appId = params[@"appId"];
    NSString *serverUrl = params[@"serverUrl"];
    NSString *debugMode = params[@"debugMode"];
    BOOL enableLog = [params[@"enableLog"] boolValue];
    NSArray *autoTrack = params[@"autoTrack"];

    if (enableLog) {
        [ThinkingAnalyticsSDK setLogLevel:TDLoggingLevelDebug];
    } else {
        [ThinkingAnalyticsSDK setLogLevel:TDLoggingLevelNone];
    }
    TDConfig *config = [TDConfig defaultTDConfig];
    config.appid = appId;
    config.configureURL = serverUrl;
    if ([debugMode isEqualToString:@"debug"]) {
        config.debugMode = ThinkingAnalyticsDebugOn;
    } else if ([debugMode isEqualToString:@"debugOnly"]) {
        config.debugMode = ThinkingAnalyticsDebugOnly;
    } else {
        config.debugMode = ThinkingAnalyticsDebugOff;
    }
    ThinkingAnalyticsAutoTrackEventType autoTrackType = ThinkingAnalyticsEventTypeNone;
    if ([autoTrack containsObject:@"appInstall"]) {
        autoTrackType = autoTrackType | ThinkingAnalyticsEventTypeAppInstall;
    }
    if ([autoTrack containsObject:@"appStart"]) {
        autoTrackType = autoTrackType | ThinkingAnalyticsEventTypeAppStart;
    }
    if ([autoTrack containsObject:@"appEnd"]) {
        autoTrackType = autoTrackType | ThinkingAnalyticsEventTypeAppEnd;
    }
    config.autoTrackEventType = autoTrackType;
    
    [ThinkingAnalyticsSDK startWithConfig:config];
}

+ (void)track:(NSDictionary *)params {
    NSString *eventName = params[@"eventName"];
    NSDictionary *properties = params[@"properties"];
    NSString *appId = params[@"appId"];
    [[ThinkingPluginProxy sharedInstance:appId] track:eventName
                                      properties:properties];
}

+ (void)trackFirst:(NSDictionary *)params {
    NSString *eventName = params[@"eventName"];
    NSString *eventId = params[@"eventId"];
    NSDictionary *properties = params[@"properties"];
    TDFirstEventModel *eventModel = [[TDFirstEventModel alloc] initWithEventName:eventName firstCheckID:eventId];
    eventModel.properties = properties;
    NSString *appId = params[@"appId"];
    [[ThinkingPluginProxy sharedInstance:appId] trackWithEventModel:eventModel];
}

+ (void)trackUpdate:(NSDictionary *)params {
    NSString *eventName = params[@"eventName"];
    NSString *eventId = params[@"eventId"];
    NSDictionary *properties = params[@"properties"];
    TDUpdateEventModel *eventModel = [[TDUpdateEventModel alloc] initWithEventName:eventName eventID:eventId];
    eventModel.properties = properties;
    NSString *appId = params[@"appId"];
    [[ThinkingPluginProxy sharedInstance:appId] trackWithEventModel:eventModel];
}

+ (void)trackOverwrite:(NSDictionary *)params {
    NSString *eventName = params[@"eventName"];
    NSString *eventId = params[@"eventId"];
    NSDictionary *properties = params[@"properties"];
    TDOverwriteEventModel *eventModel = [[TDOverwriteEventModel alloc] initWithEventName:eventName eventID:eventId];
    eventModel.properties = properties;
    NSString *appId = params[@"appId"];
    [[ThinkingPluginProxy sharedInstance:appId] trackWithEventModel:eventModel];
}

+ (void)timeEvent:(NSDictionary *)params {
    NSString *eventName = params[@"eventName"];
    NSString *appId = params[@"appId"];
    [[ThinkingPluginProxy sharedInstance:appId] timeEvent:eventName];
}

+ (void)userSet:(NSDictionary *)params {
    NSDictionary *properties = params[@"properties"];
    NSString *appId = params[@"appId"];
    [[ThinkingPluginProxy sharedInstance:appId] user_set:properties];
}

+ (void)userSetOnce:(NSDictionary *)params {
    NSDictionary *properties = params[@"properties"];
    NSString *appId = params[@"appId"];
    [[ThinkingPluginProxy sharedInstance:appId] user_setOnce:properties];
}

+ (void)userUnset:(NSDictionary *)params {
    NSString *property = params[@"property"];
    NSString *appId = params[@"appId"];
    [[ThinkingPluginProxy sharedInstance:appId] user_unset:property];
}

+ (void)userAdd:(NSDictionary *)params {
    NSDictionary *properties = params[@"properties"];
    NSString *appId = params[@"appId"];
    [[ThinkingPluginProxy sharedInstance:appId] user_add:properties];
}

+ (void)userAppend:(NSDictionary *)params {
    NSDictionary *properties = params[@"properties"];
    NSString *appId = params[@"appId"];
    [[ThinkingPluginProxy sharedInstance:appId] user_append:properties];
}

+ (void)userUniqAppend:(NSDictionary *)params {
    NSDictionary *properties = params[@"properties"];
    NSString *appId = params[@"appId"];
    [[ThinkingPluginProxy sharedInstance:appId] user_uniqAppend:properties];
}

+ (void)userDel:(NSDictionary *)params {
    NSString *appId = params[@"appId"];
    [[ThinkingPluginProxy sharedInstance:appId] user_delete];
}

+ (void)flush:(NSDictionary *)params {
    NSString *appId = params[@"appId"];
    [[ThinkingPluginProxy sharedInstance:appId] flush];
}

+ (void)login:(NSDictionary *)params {
    NSString *accountId = params[@"accountId"];
    NSString *appId = params[@"appId"];
    [[ThinkingPluginProxy sharedInstance:appId] login:accountId];
}

+ (void)logout:(NSDictionary *)params {
    NSString *appId = params[@"appId"];
    [[ThinkingPluginProxy sharedInstance:appId] logout];
}

+ (void)identify:(NSDictionary *)params {
    NSString *distinctId = params[@"distinctId"];
    NSString *appId = params[@"appId"];
    [[ThinkingPluginProxy sharedInstance:appId] identify:distinctId];
}

+ (NSDictionary *)getDistinctId:(NSDictionary *)params {
    NSString *appId = params[@"appId"];
    NSString *distinctId = [[ThinkingPluginProxy sharedInstance:appId] getDistinctId];
    NSDictionary *ret = @{
        @"message": @{ @"distinctId": distinctId },
        @"type": @"getDistinctId"
    };
    return ret;
}

+ (NSDictionary *)getDeviceId:(NSDictionary *)params {
    NSString *appId = params[@"appId"];
    NSString *deviceId = [[ThinkingPluginProxy sharedInstance:appId] getDeviceId];
    NSDictionary *ret = @{
        @"message": @{ @"deviceId": deviceId },
        @"type": @"getDeviceId"
    };
    return ret;
}

+ (void)setSuperProperties:(NSDictionary *)params {
    NSDictionary *properties = params[@"properties"];
    NSString *appId = params[@"appId"];
    [[ThinkingPluginProxy sharedInstance:appId] setSuperProperties:properties];
}

+ (void)unsetSuperProperties:(NSDictionary *)params {
    NSString *property = params[@"property"];
    NSString *appId = params[@"appId"];
    [[ThinkingPluginProxy sharedInstance:appId] unsetSuperProperty:property];
}

+ (NSDictionary *)getSuperProperties:(NSDictionary *)params {
    NSString *appId = params[@"appId"];
    NSDictionary *properties = [[ThinkingPluginProxy sharedInstance:appId] currentSuperProperties];
    NSDictionary *ret = @{
        @"message": @{ @"properties": properties },
        @"type": @"getSuperProperties"
    };
    return ret;
}

+ (void)clearSuperProperties:(NSDictionary *)params {
    NSString *appId = params[@"appId"];
    [[ThinkingPluginProxy sharedInstance:appId] clearSuperProperties];
}

+ (NSDictionary *)getPresetProperties:(NSDictionary *)params {
    NSString *appId = params[@"appId"];
    NSDictionary *properties = [[[ThinkingPluginProxy sharedInstance:appId] getPresetProperties] toEventPresetProperties];
    NSDictionary *ret = @{
        @"message": @{ @"properties": properties },
        @"type": @"getPresetProperties"
    };
    return ret;
}

+ (void)setCustomerLibInfo:(NSDictionary *)params {
    NSString *libName = params[@"libName"];
    NSString *libVersion = params[@"libVersion"];
    [ThinkingAnalyticsSDK setCustomerLibInfoWithLibName:libName libVersion:libVersion];
}

+ (void)calibrateTime:(NSDictionary *)params {
    NSTimeInterval timestamp = [params[@"timestamp"] doubleValue];
    [ThinkingAnalyticsSDK calibrateTime:timestamp];
}

+ (void)calibrateTimeWithNtp:(NSDictionary *)params {
    NSString *ntpServer = params[@"ntpServer"];
    [ThinkingAnalyticsSDK calibrateTimeWithNtp:ntpServer];
}

@end

File: android/plugin/src/main/java/plugin/library/ThinkingPluginProxy.java
package plugin.library;

import android.content.Context;
import android.text.TextUtils;

import com.ansca.corona.CoronaEnvironment;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

import cn.thinkingdata.analytics.TDConfig;
import cn.thinkingdata.analytics.TDFirstEvent;
import cn.thinkingdata.analytics.TDOverWritableEvent;
import cn.thinkingdata.analytics.TDUpdatableEvent;
import cn.thinkingdata.analytics.ThinkingAnalyticsSDK;

public class ThinkingPluginProxy {

    private static Context ceContext = CoronaEnvironment.getApplicationContext();

    private static String defaultAppId;

    public static void setCustomerLibInfo(JSONObject params) {
        String libName = params.optString("libName");
        String libVersion = params.optString("libVersion");
        ThinkingAnalyticsSDK.setCustomerLibInfo(libName, libVersion);
    }

    public static void shareInstance(JSONObject params) {
        String appId = params.optString("appId");
        String serverUrl = params.optString("serverUrl");
        String debugMode = params.optString("debugMode");
        boolean enableLog = params.optBoolean("enableLog");
        JSONArray autoTrack = params.optJSONArray("autoTrack");

        if (TextUtils.isEmpty(defaultAppId)) {
            defaultAppId = appId;
        }

        ThinkingAnalyticsSDK.enableTrackLog(enableLog);

        TDConfig config = TDConfig.getInstance(ceContext, appId, serverUrl);
        if (debugMode.equals("debug")) {
            config.setMode(TDConfig.ModeEnum.DEBUG);
        } else if (debugMode.equals("debugOnly")) {
            config.setMode(TDConfig.ModeEnum.DEBUG_ONLY);
        } else  {
            config.setMode(TDConfig.ModeEnum.NORMAL);
        }
        List<ThinkingAnalyticsSDK.AutoTrackEventType> autoTrackType = new ArrayList<>();
        for(int i=0; autoTrack!=null&&i<autoTrack.length(); i++) {
            String type = autoTrack.optString(i);
            if ("appInstall".equals(type)) {
                autoTrackType.add(ThinkingAnalyticsSDK.AutoTrackEventType.APP_INSTALL);
            } else if ("appStart".equals(type)) {
                autoTrackType.add(ThinkingAnalyticsSDK.AutoTrackEventType.APP_START);
            } else if ("appEnd".equals(type)) {
                autoTrackType.add(ThinkingAnalyticsSDK.AutoTrackEventType.APP_END);
            }
        }
        ThinkingAnalyticsSDK.sharedInstance(config);
        ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).enableAutoTrack(autoTrackType);
    }

    public static void track(JSONObject params) {
        String appId = params.optString("appId");
        String eventName = params.optString("eventName");
        JSONObject properties = params.optJSONObject("properties");
        if (TextUtils.isEmpty(appId)) {
            appId = defaultAppId;
        }
        ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).track(eventName, properties);
    }

    public static void trackFirst(JSONObject params) {
        String appId = params.optString("appId");
        String eventName = params.optString("eventName");
        String eventId = params.optString("eventId");
        JSONObject properties = params.optJSONObject("properties");
        TDFirstEvent firstEvent = new TDFirstEvent(eventName, properties);
        if (eventId != null) {
            firstEvent.setFirstCheckId(eventId);
        }
        if (TextUtils.isEmpty(appId)) {
            appId = defaultAppId;
        }
        ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).track(firstEvent);
    }

    public static void trackUpdate(JSONObject params) {
        String appId = params.optString("appId");
        String eventName = params.optString("eventName");
        String eventId = params.optString("eventId");
        JSONObject properties = params.optJSONObject("properties");
        TDUpdatableEvent updatableEvent = new TDUpdatableEvent(eventName, properties, eventId);
        if (TextUtils.isEmpty(appId)) {
            appId = defaultAppId;
        }
        ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).track(updatableEvent);
    }

    public static void trackOverwrite(JSONObject params) {
        String appId = params.optString("appId");
        String eventName = params.optString("eventName");
        String eventId = params.optString("eventId");
        JSONObject properties = params.optJSONObject("properties");
        TDOverWritableEvent overWritableEvent = new TDOverWritableEvent(eventName, properties, eventId);
        if (TextUtils.isEmpty(appId)) {
            appId = defaultAppId;
        }
        ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).track(overWritableEvent);
    }

    public static void timeEvent(JSONObject params) {
        String appId = params.optString("appId");
        String eventName = params.optString("eventName");
        if (TextUtils.isEmpty(appId)) {
            appId = defaultAppId;
        }
        ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).timeEvent(eventName);
    }

    public static void userSet(JSONObject params) {
        String appId = params.optString("appId");
        JSONObject properties = params.optJSONObject("properties");
        if (TextUtils.isEmpty(appId)) {
            appId = defaultAppId;
        }
        ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).user_set(properties);
    }

    public static void userSetOnce(JSONObject params) {
        String appId = params.optString("appId");
        JSONObject properties = params.optJSONObject("properties");
        if (TextUtils.isEmpty(appId)) {
            appId = defaultAppId;
        }
        ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).user_setOnce(properties);
    }

    public static void userUnset(JSONObject params) {
        String appId = params.optString("appId");
        String property = params.optString("property");
        if (TextUtils.isEmpty(appId)) {
            appId = defaultAppId;
        }
        ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).user_unset(property);
    }

    public static void userAdd(JSONObject params) {
        String appId = params.optString("appId");
        JSONObject properties = params.optJSONObject("properties");
        if (TextUtils.isEmpty(appId)) {
            appId = defaultAppId;
        }
        ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).user_add(properties);
    }

    public static void userAppend(JSONObject params) {
        String appId = params.optString("appId");
        JSONObject properties = params.optJSONObject("properties");
        if (TextUtils.isEmpty(appId)) {
            appId = defaultAppId;
        }
        ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).user_append(properties);
    }

    public static void userUniqAppend(JSONObject params) {
        String appId = params.optString("appId");
        JSONObject properties = params.optJSONObject("properties");
        if (TextUtils.isEmpty(appId)) {
            appId = defaultAppId;
        }
        ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).user_uniqAppend(properties);
    }

    public static void userDel(JSONObject params) {
        String appId = params.optString("appId");
        if (TextUtils.isEmpty(appId)) {
            appId = defaultAppId;
        }
        ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).user_delete();
    }

    public static void flush(JSONObject params) {
        String appId = params.optString("appId");
        if (TextUtils.isEmpty(appId)) {
            appId = defaultAppId;
        }
        ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).flush();
    }

    public static void login(JSONObject params) {
        String appId = params.optString("appId");
        String accountId = params.optString("accountId");
        if (TextUtils.isEmpty(appId)) {
            appId = defaultAppId;
        }
        ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).login(accountId);
    }

    public static void logout(JSONObject params) {
        String appId = params.optString("appId");
        if (TextUtils.isEmpty(appId)) {
            appId = defaultAppId;
        }
        ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).logout();
    }

    public static void identify(JSONObject params) {
        String appId = params.optString("appId");
        String distinctId = params.optString("distinctId");
        if (TextUtils.isEmpty(appId)) {
            appId = defaultAppId;
        }
        ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).identify(distinctId);
    }

    public static JSONObject getDistinctId(JSONObject params) {
        String appId = params.optString("appId");
        if (TextUtils.isEmpty(appId)) {
            appId = defaultAppId;
        }
        String distinctId = ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).getDistinctId();
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put("type", "getDistinctId");
            jsonObject.put("message", new JSONObject().put("distinctId", distinctId));
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return jsonObject;
    }

    public static JSONObject getDeviceId(JSONObject params) {
        String appId = params.optString("appId");
        if (TextUtils.isEmpty(appId)) {
            appId = defaultAppId;
        }
        String deviceId = ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).getDeviceId();
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put("type", "getDeviceId");
            jsonObject.put("message", new JSONObject().put("deviceId", deviceId));
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return jsonObject;
    }

    public static void setSuperProperties(JSONObject params) {
        String appId = params.optString("appId");
        JSONObject properties = params.optJSONObject("properties");
        if (TextUtils.isEmpty(appId)) {
            appId = defaultAppId;
        }
        ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).setSuperProperties(properties);
    }

    public static void unsetSuperProperties(JSONObject params) {
        String appId = params.optString("appId");
        String property = params.optString("property");
        if (TextUtils.isEmpty(appId)) {
            appId = defaultAppId;
        }
        ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).unsetSuperProperty(property);
    }

    public static JSONObject getSuperProperties(JSONObject params) {
        String appId = params.optString("appId");
        if (TextUtils.isEmpty(appId)) {
            appId = defaultAppId;
        }
        JSONObject properties = ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).getSuperProperties();
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put("type", "getSuperProperties");
            jsonObject.put("message", new JSONObject().put("properties", properties));
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return jsonObject;
    }

    public static void clearSuperProperties(JSONObject params) {
        String appId = params.optString("appId");
        if (TextUtils.isEmpty(appId)) {
            appId = defaultAppId;
        }
        ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).clearSuperProperties();
    }

    public static JSONObject getPresetProperties(JSONObject params) {
        String appId = params.optString("appId");
        if (TextUtils.isEmpty(appId)) {
            appId = defaultAppId;
        }
        JSONObject properties = ThinkingAnalyticsSDK.sharedInstance(ceContext, appId).getPresetProperties().toEventPresetProperties();
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put("type", "getPresetProperties");
            jsonObject.put("message", new JSONObject().put("properties", properties));
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return jsonObject;
    }

    public static void calibrateTime(JSONObject params) {
        long timestamp = params.optLong("timestamp");
        ThinkingAnalyticsSDK.calibrateTime(timestamp);
    }

    public static void calibrateTimeWithNtp(JSONObject params) {
        String ntpServer = params.optString("ntpServer");
        ThinkingAnalyticsSDK.calibrateTimeWithNtp(ntpServer);
    }
}
